037-pluggable-pod-security-profiles.md

```diff
@@ -0,0 +1,234 @@
+
+The pluggable mechanism should also make it easy to add additional implementations to Strimzi in the future.
+
+### Plugin implementation
+
+#### Interface
+
+A new interface `PodSecurityProvider` will be added to the `api` module.
+All provider plugins will implement this interface.
+The reason for having this interface in the `api` module is that this module is distributed though Maven repositories.
+Users will be able to write their implementations by using the `api` module as dependency.
+
+The interface will have methods for providing security contexts for all Pods created by Strimzi, and all containers.
+The methods providing the security context will get as a parameter an object of type `PodSecurityProviderContext`.
+This will be an object created by Strimzi which will encapsulate the different infromation which strimzi will provide to the Provider plugin to make its decisions when generating the Kubernetes security context.
+The `PodSecurityProviderContext` will be defines as another interface in the `api` module:
+
+```java
+interface PodSecurityProviderContext {
+    Storage storage();
+    PodSecurityContext userSuppliedContext();
+}
+```
+
+This encapsulation should make it easier to add additional information in the future without changing the signature of the provider methods but only by adding new methods to the Strimzi provided context.
+
+The provider methods will have the default implementation which will always just return the user-specified context.
+For example (just few methods are listed in this example):
+
+```java
+public interface PodSecurityProvider {
+    // ...
+
+    default PodSecurityContext kafkaPodSecurityContext(PodSecurityProviderContext context) {
+        return context.userSuppliedContext();
+    }
+
+    default SecurityContext kafkaContainerSecurityContext(PodSecurityProviderContext context) {
+        return context.userSuppliedContext();
+    }
+
+    default SecurityContext kafkaInitContainerSecurityContext(PodSecurityProviderContext context) {
+        return context.userSuppliedContext();
+    }
+
+    // ...
+}
+```
+
+The interface will also have a `configure` method for configuring the provider.
+The `configure` method will get as a parameter an object describing the platform (Kubernetes version, features / APIs) which can be taken into account when creating the security context.
+This method will not have any default implementation.
+
+```java
+public interface PodSecurityProvider {
+    // ...
+    void configure(PlatformFeatures platformFeatures);
+    // ...
+}
+```
+
+`PlatformFeatures` in this method is another new interface which provides some of the methods provided by the `PlatformFeaturesAvailability` class which is part of the `operator-common` module`
+It uses the features which might be relevant for generating the securitny context:
+
+```java
+public interface PlatformFeatures {
+    KubernetesVersion getKubernetesVersion();
+    boolean isOpenshift();
```

Comment:
Would something like `platformSuppliesDefaultContext()` actually capture the behaviour we're after? That way the API could support other kube distributions doing similar things to OpenShift, without having to represent every possible distribution in the API.

Reply:
TBH, I don't know. I'm not sure what other platforms are doing -> some other platforms might simply validate the context without injecting it. Even with OpenShift - it currently injects the security context. But not sure how it will do it in the future once it migrates from SCC to Kubernetes Pod Security Profiles. So I'm not entirely sure if something like `platformSuppliesDefaultContext` makes sense as it might be equally insufficient in the future as `isOpenshift` is. It also shifts more of the logic from the plugin into the core operator which might need to identify the platform and decide what it does. With the currently proposed API, you can do more directly in the plugin - for example decide for different behaviour on OpenShift with version A.B.x versus OpenShift with version C.D.x. So I would probably be inclined to keep the `isOpenshift` as proposed right now.